### PR TITLE
Fix ESCs constantly arming on rover with dshot commands

### PR DESCRIPTION
**Pull request by andyp1per**
Several bugs fixed:

- ESCs constantly arming on rover with dshot commands
- Bdshot not working with fewer channels than the full group
- Not disabling disabled channels
- Not pulling high all channels in a bdshot group (caused random motor movements in disabled channels)
- Not ignoring disabled channels for telemetry starting position

Although this affects all vehicles it will particularly help rover and plane

### CODE REVIEW

File: libraries/SRV_Channel/SRV_Channel.h [hunk 1]
```diff
@@ -183,7 +183,7 @@ class SRV_Channel {
 
     // check if a function is valid for indexing into functions
     static bool valid_function(Aux_servo_function_t fn) {
-        return fn >= 0 && fn < k_nr_aux_servo_functions;
+        return fn > k_none && fn < k_nr_aux_servo_functions;
```
Comment (IamPete1) [inline libraries/SRV_Channel/SRV_Channel.h:186]: This looks like a drive by? I fear un-intended consequences. 

I think it means that setting servos to none at runtime will no longer stop there PWM output.

Comment (andyp1per) [reply libraries/SRV_Channel/SRV_Channel.h:186]: where is the output currrently stopped?

Comment (IamPete1) [reply libraries/SRV_Channel/SRV_Channel.h:186]: https://github.com/ArduPilot/ardupilot/blob/1521760a1713a4911956d6c6e79e296be751fd4e/libraries/SRV_Channel/SRV_Channels.cpp#L330-L332

I have not tested of course....

Maybe we could just special case none here:

https://github.com/ArduPilot/ardupilot/blob/1521760a1713a4911956d6c6e79e296be751fd4e/libraries/SRV_Channel/SRV_Channel_aux.cpp#L213-L224

Although the PWM channel would now not be enabled a few lines above. 

Which mask do you need to be correct?

Comment (andyp1per) [reply libraries/SRV_Channel/SRV_Channel.h:186]: Motor enabled masks

Comment (andyp1per) [reply libraries/SRV_Channel/SRV_Channel.h:186]: What about:
```
        if (channels[i].valid_function()) {
            channels[i].calc_pwm(functions[channels[i].function.get()].output_scaled);
        } else {
            channels[i].calc_pwm(0.0f);
        }
```
?

Comment (andyp1per) [reply libraries/SRV_Channel/SRV_Channel.h:186]: Surely GPIO currently has the same issue?

Comment (andyp1per) [reply libraries/SRV_Channel/SRV_Channel.h:186]: Actually I have a better fix that fixes a pre-existing issue

Comment (andyp1per) [reply libraries/SRV_Channel/SRV_Channel.h:186]: ... and the fix uncovers a bug in SITL

File: libraries/AP_HAL_ChibiOS/RCOutput.cpp [hunk 2]
```diff
@@ -1401,15 +1412,11 @@ void RCOutput::dshot_send(pwm_group &group, uint32_t time_out_us)
...
 
             if (safety_on && !(safety_mask & (1U<<(chan+chan_offset)))) {
-                // safety is on, overwride pwm
-                pwm = 0;
+                // safety is on, override pwm
+                continue;
             }
 
             const uint16_t chan_mask = (1U<<chan);
-            if (pwm == 0) {
```
Comment (IamPete1) [inline libraries/AP_HAL_ChibiOS/RCOutput.cpp:1409]: pwm could be 0 for reasons other than the safety mask.

Comment (andyp1per) [reply libraries/AP_HAL_ChibiOS/RCOutput.cpp:1409]: Setting pwm to 0 is not the right way to disable dshot. I can probably put this check back in, but it's not right.

**Comment by rmackay9** [comment]
I've tested this on a pixracer (using the special bdshot firmware for Rover) and confirmed that it works!
image

With master the ESC does not arm, with this PR it spins correctly.

**Comment by andyp1per** [comment]
Thanks @rmackay9 should be fixed now. There were actually two remaining bugs, both of which were extremely subtle.

File: libraries/AP_HAL_ChibiOS/RCOutput.cpp (hunk 2 above)
Comment (andyp1per) [reply libraries/AP_HAL_ChibiOS/RCOutput.cpp:1409]: OK I have put it back in

**Comment by andyp1per** [comment]
Flown on a miniquad without issue - bi-directional dshot all ok.

**Review by rmackay9 — APPROVED** [review]
I've tested this and it resolves the issues I was seeing on Rover, thanks very much!

**Comment by rmackay9** [comment]
This has been added to Copter-4.2.0-beta3

**Pull request merged**